revoke.go

```diff
@@ -0,0 +1,81 @@
+package main
+
+import (
+	"context"
+	"fmt"
+	"github.com/coreos/go-oidc"
+	"github.com/pkg/errors"
+	"golang.org/x/oauth2"
+	"io/ioutil"
+	"net/http"
+	"net/url"
+	"strings"
+)
+
+// revocationEndpoint parses the OIDC Provider claims from the discovery document
+// and tries to find the revocation_endpoint.
+func revocationEndpoint(p *oidc.Provider) (string, error) {
+	claims := struct {
+		RevocationEndpoint string `json:"revocation_endpoint"`
+	}{}
+	if err := p.Claims(&claims); err != nil {
+		return "", errors.Wrap(err, "Error unmarshalling provider doc into struct")
+	}
+	if claims.RevocationEndpoint == "" {
+		return "", errors.New("Provider doesn't have a revocation_endpoint")
+	}
+	return claims.RevocationEndpoint, nil
+}
+
+// revokeTokens is a helper that takes an oauth2.Token and revokes the access and refresh tokens.
+// If no tokens are found, it succeeds.
+func revokeTokens(ctx context.Context, revocationEndpoint string, token *oauth2.Token, clientID, clientSecret string) error {
+	if token.AccessToken != "" {
+		err := revokeToken(ctx, revocationEndpoint, token.AccessToken, "access_token", clientID, clientSecret)
+		if err != nil {
+			return errors.Wrap(err, "Failed to revoke access token")
+		}
+	}
+	if token.RefreshToken != "" {
+		err := revokeToken(ctx, revocationEndpoint, token.AccessToken, "refresh_token", clientID, clientSecret)
+		if err != nil {
+			return errors.Wrap(err, "Failed to revoke refresh token")
+		}
+	}
+	return nil
+}
+
+// revokeToken takes care of revoking an access/refresh token to the IdP.
+// The revocation procedure is described in RFC7009:+
+// https://tools.ietf.org/html/rfc7009
+func revokeToken(ctx context.Context, revocationEndpoint string, token, tokenType, clientID, clientSecret string) error {
+	// Verify revocation_endpoint has https url
+	if !strings.HasPrefix(revocationEndpoint, "https") {
+		return errors.New(fmt.Sprintf("Revocation endpoint (%v) MUST use https", revocationEndpoint))
+	}
+	values := url.Values{}
+	values.Set("token", token)
+	values.Set("token_type_hint", tokenType)
+	req, err := http.NewRequest(http.MethodPost, revocationEndpoint, strings.NewReader(values.Encode()))
+	if err != nil {
+		return err
+	}
+	req.Header.Set("Content-Type", "application/x-www-form-urlencoded")
+	// Assume provider accepts http basic auth, because it is REQUIRED by the spec
+	req.SetBasicAuth(clientID, clientSecret)
+
+	resp, err := doRequest(ctx, req)
+	if err != nil {
+		return errors.Wrap(err, "Error contacting revocation endpoint")
+	}
+	if code := resp.StatusCode; code != 200 {
+		// Try to parse JSON response
+		defer resp.Body.Close()
+		body, err := ioutil.ReadAll(resp.Body)
+		if err != nil {
+			return errors.New(fmt.Sprintf("Revocation endpoint returned code %v, failed to read body: %v", code, err))
+		}
+		return errors.New(fmt.Sprintf("Revocation endpoint returned code %v, server returned: %v", code, body))
```

Comment:
I think there are two issues here:

1. We can't be sure that the HTTP response is in JSON format. An intermediate proxy server can just as well send an HTML response. So, it makes sense to print it as well, without trying to interpret the body. That's roughly what the spec says as well:

   >    The content of the response body is ignored by the client as all
   >    necessary information is conveyed in the response code.

2. I'd probably (but I'm not dead set on this) limit the number of characters that may be printed, which is especially necessary for authorization servers with long HTML error pages.

Reply:
@apyrgio [RFC7079 section on revocation](https://tools.ietf.org/html/rfc7009#section-2.2.1) response says:

> The error presentation conforms to the definition in Section 5.2 of
   [RFC6749].

And [RFC6749](https://tools.ietf.org/html/rfc6749#section-5.2) says:

> The parameters are included in the entity-body of the HTTP response
   using the "application/json" media type as defined by [RFC4627].  The
   parameters are serialized into a JSON structure by adding each
   parameter at the highest structure level.

Did I miss something? Doesn't the spec require the response to be json?

Reply:
Oh, you're right, I somehow skipped this part in the RFC. Still, my original point is that in case of an error, you can't be sure that it's the server that returns the error response. It may as well be Nginx, or the stored endpoint may be incorrect, and thus a third-party may respond. In these cases, we still want to show the error response for debug purposes, even if it's not in JSON format.

Reply:
I see that I wasn't doing any JSON parsing anyway, so I changed the comment to mention that if we want to parse this someday, it will probably be JSON.
Also, about this:
> It may as well be Nginx, or the stored endpoint may be incorrect, and thus a third-party may respond.

Are you saying that Nginx would change the content-type and the body of the response?
That seems illogical to me.

Reply:
> I see that I wasn't doing any JSON parsing anyway, so I changed the comment to mention that if we want to parse this someday, it will probably be JSON.

I was about to ask that, but I though it was somehow handled internally in `ioutil.ReadAll()`.

> Are you saying that Nginx would change the content-type and the body of the response?
That seems illogical to me.

I'm saying that the authorization server may not even have the chance to receive the request. It may error out before it reaches it, e.g., due to a misconfiguration or a temporary error. In this case, all bets are off for the type of the error response. And that's highly likely, because we're dealing with a third-party endpoint here.

In any case, as long as we show the body (which we do now), I think we're ok.